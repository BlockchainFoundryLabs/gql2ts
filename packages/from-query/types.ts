import {
  OperationTypeNode,
  SelectionNode,
  GraphQLType,
  DefinitionNode,
  NamedTypeNode,
  GraphQLNamedType,
  GraphQLOutputType,
  GraphQLInputType,
  TypeNode,
  OperationDefinitionNode,
} from 'graphql';
import { PossibleSchemaInput } from '@gql2ts/util';

export type ChildSelectionsType =
  (operation: OperationTypeNode, selection: SelectionNode, indentation: string, parent?: GraphQLType, isUndefined?: boolean)
    => IChildren;

export interface IProvidedOptions extends Partial<IOptions> {};

export type Signature = (schema: PossibleSchemaInput, query: string, typeMap?: Partial<ITypeMap>, options?: IProvidedOptions) => IReturn[];

export interface IComplexTypeSignature {
  iface: string;
  isPartial: boolean;
  name: string;
}
export interface IChildren {
  isFragment: boolean;
  isPartial: boolean;
  iface: string;
  complexTypes: IComplexTypeSignature[];
}

export interface IReturn {
  variables: string;
  interface: string;
  additionalTypes: string[];
}

export type BuildRootInterfaceName = (definition: DefinitionNode, queryNamer: QueryNamer, fragmentNamer: WrapType) => string;
export type InterfaceFormatters = (operationName: string, fields: string[]) => string;
export type FragmentInterfaceFormatter = (operationName: string, fields: string[], interfaceExtensions: string[]) => string;
export type GenerateSubTypeInterface = (selectionName: string, generatedCount: number) => string;
export type WrapType = (type: string) => string;
export type TypePrinter = (type: string, isNonNull: boolean) => string;
export type InputFormatter = (name: string, isOptional: boolean, type: string) => string;
export type QueryNamer = (def: OperationDefinitionNode) => string;

export interface IOptions {
  buildRootInterfaceName: BuildRootInterfaceName;
  formatVariableInterface: InterfaceFormatters;
  formatInterface: InterfaceFormatters;
  formatFragmentInterface: FragmentInterfaceFormatter;
  wrapList: WrapType;
  wrapPartial: WrapType;
  generateSubTypeInterfaceName: GenerateSubTypeInterface;
  printType: TypePrinter;
  formatInput: InputFormatter;
  generateFragmentName: WrapType;
  generateQueryName: QueryNamer;
};

export type RegularTypeSignature = (type: NamedTypeNode | GraphQLNamedType, isNonNull: boolean, replacement: string | null) => string;
export type VariableTypeSignature = (type: TypeNode, isNonNull?: boolean, replacement?: string | null) => string;
export type convertToTypeSignature =
  (type: GraphQLOutputType | GraphQLInputType, isNonNull?: boolean, replacement?: string | null) => string;

export interface IDefaultTypeMap {
  ID: string;
  String: string;
  Boolean: string;
  Float: string;
  Int: string;
  __DEFAULT: string;
}

export interface ITypeMap extends IDefaultTypeMap {
  [x: string]: string | undefined;
}
