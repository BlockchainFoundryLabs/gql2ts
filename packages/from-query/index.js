"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const graphql_1 = require("graphql");
const util_1 = require("@gql2ts/util");
const DEFAULT_FORMAT_INTERFACE = (opName, fields) => `export interface ${opName} {
${fields.join('\n  ')}
}`;
const DEFAULT_FORMAT_VARIABLES = (opName, fields) => `export interface ${opName}Input {
  ${fields.join('\n  ')}
}`;
const DEFAULT_FORMAT_FRAGMENT = (opName, fields, ext) => `export interface ${opName}${ext} {
${fields.join('\n')}
}`;
const DEFAULT_BUILD_ROOT_INTERFACE_NAME = def => {
    if (def.kind === 'OperationDefinition') {
        return def.name ? def.name.value : 'Anonymous';
    }
    else if (def.kind === 'FragmentDefinition') {
        return `IFragment${def.name.value}`;
    }
    else {
        throw new Error(`Unsupported Definition ${def.kind}`);
    }
};
const DEFAULT_TYPE_MAP = {
    ID: 'string',
    String: 'string',
    Boolean: 'boolean',
    Float: 'number',
    Int: 'number',
};
const DEFAULT_WRAP_LIST = type => `Array<${type}>`;
const DEFAULT_WRAP_PARTIAL = possiblePartial => {
    if (possiblePartial.isPartial) {
        return `Partial<${possiblePartial.iface}>`;
    }
    else {
        return possiblePartial.iface;
    }
};
const DEFAULT_GENERATE_SUBTYPE_INTERFACE_NAME = (selectionName, generatedCount) => `SelectionOn${selectionName}${!!generatedCount ? generatedCount : ''}`;
;
const DEFAULT_OPTIONS = {
    buildRootInterfaceName: DEFAULT_BUILD_ROOT_INTERFACE_NAME,
    formatVariableInterface: DEFAULT_FORMAT_VARIABLES,
    formatInterface: DEFAULT_FORMAT_INTERFACE,
    formatFragmentInterface: DEFAULT_FORMAT_FRAGMENT,
    wrapList: DEFAULT_WRAP_LIST,
    wrapPartial: DEFAULT_WRAP_PARTIAL,
    generateSubTypeInterfaceName: DEFAULT_GENERATE_SUBTYPE_INTERFACE_NAME,
};
;
const doIt = (schema, query, typeMap = {}, providedOptions = {}) => {
    const TypeMap = Object.assign({}, DEFAULT_TYPE_MAP, typeMap);
    const options = Object.assign({}, DEFAULT_OPTIONS, providedOptions);
    const { buildRootInterfaceName, formatFragmentInterface, formatInterface, formatVariableInterface, wrapList, wrapPartial, generateSubTypeInterfaceName, } = options;
    const parsedSchema = util_1.schemaFromInputs(schema);
    const parsedSelection = graphql_1.parse(query);
    function isNonNullable(type) {
        return type instanceof graphql_1.GraphQLNonNull;
    }
    function isList(type) {
        return type instanceof graphql_1.GraphQLList;
    }
    const printType = (type, isNonNull) => isNonNull ? type : `${type} | null`;
    const handleInputObject = (type, isNonNull) => {
        const variables = Object.keys(type.getFields()).map(k => type.getFields()[k]);
        const builder = `{\n${variables.map(v => `    ${v.name}?: ${convertToType(v.type)};`).join('\n')}\n  }`;
        return printType(builder, isNonNull);
    };
    const handleEnum = (type, isNonNull) => {
        const decl = type.getValues().map(en => `'${en.value}'`).join(' | ');
        return printType(decl, isNonNull);
    };
    const handleNamedTypeInput = (type, isNonNull) => {
        if (type.kind === 'NamedType' && type.name.kind === 'Name' && type.name.value) {
            const newType = parsedSchema.getType(type.name.value);
            if (newType instanceof graphql_1.GraphQLEnumType) {
                return handleEnum(newType, isNonNull);
            }
            else if (newType instanceof graphql_1.GraphQLInputObjectType) {
                return handleInputObject(newType, isNonNull);
            }
        }
    };
    const handleRegularType = (type, isNonNull, replacement) => {
        const typeValue = (typeof type.name === 'string') ? type.toString() : type.name.value;
        const showValue = replacement ? replacement : typeValue;
        const show = TypeMap[showValue] || (replacement ? showValue : 'any');
        return printType(show, isNonNull);
    };
    const convertVariable = (type, isNonNull = false, replacement = null) => {
        if (type.kind === 'ListType') {
            return wrapList(convertVariable(type.type, false, replacement)) + printType('', isNonNull);
        }
        else if (type.kind === 'NonNullType') {
            return convertVariable(type.type, true, replacement);
        }
        else {
            return handleNamedTypeInput(type, isNonNull) || handleRegularType(type, isNonNull, replacement);
        }
    };
    const convertToType = (type, isNonNull = false, replacement = null) => {
        if (isList(type)) {
            return wrapList(convertToType(type.ofType, false, replacement)) + printType('', isNonNull);
        }
        else if (isNonNullable(type)) {
            return convertToType(type.ofType, true, replacement);
        }
        else if (type instanceof graphql_1.GraphQLEnumType) {
            return handleEnum(type, isNonNull);
        }
        else {
            return handleRegularType(type, isNonNull, replacement);
        }
    };
    const UndefinedDirectives = new Set(['include', 'skip']);
    const isUndefinedFromDirective = directives => {
        if (!directives || !directives.length) {
            return false;
        }
        const badDirectives = directives.filter(d => !UndefinedDirectives.has(d.name.value));
        const hasDirectives = directives.some(d => UndefinedDirectives.has(d.name.value));
        if (badDirectives.length) {
            console.error('Found some unknown directives:');
            badDirectives.forEach(d => console.error(d.name.value));
        }
        if (hasDirectives) {
            return true;
        }
        else {
            return false;
        }
    };
    const getOperationFields = operation => {
        switch (operation) {
            case 'query':
                return parsedSchema.getQueryType();
            case 'mutation':
                return parsedSchema.getMutationType();
            case 'subscription':
                return parsedSchema.getSubscriptionType();
            default:
                throw new Error('Unsupported Operation');
        }
    };
    const getChildSelections = (operation, selection, indentation = '', parent, isUndefined = false) => {
        let str = '';
        let field;
        let isFragment = false;
        let isPartial = false;
        let generatedTypeCount = 0;
        let complexTypes = [];
        if (selection.kind === 'Field') {
            if (parent && graphql_1.isCompositeType(parent)) {
                if (parent instanceof graphql_1.GraphQLUnionType) {
                    field = parent.getTypes().map(t => t.getFields()[selection.name.value]).find(z => !!z);
                }
                else {
                    field = parent.getFields()[selection.name.value];
                }
            }
            else {
                const operationFields = getOperationFields(operation);
                field = operationFields.getFields()[selection.name.value];
            }
            let selectionName = selection.name.value;
            if (selection.alias) {
                selectionName = selection.alias.value;
            }
            if (isUndefined || isUndefinedFromDirective(selection.directives)) {
                selectionName += '?';
            }
            str += indentation + selectionName + ': ';
            if (!!selection.selectionSet) {
                let newParent;
                if (!field) {
                    console.log(selection, newParent);
                }
                const fieldType = graphql_1.getNamedType(field.type);
                if (graphql_1.isCompositeType(fieldType)) {
                    newParent = fieldType;
                }
                const selections = selection.selectionSet.selections.map(sel => getChildSelections(operation, sel, indentation + '  ', newParent));
                const nonFragments = selections.filter(s => !s.isFragment);
                const fragments = selections.filter(s => s.isFragment);
                const andOps = [];
                complexTypes.push(...selections.map(sel => sel.complexTypes).reduce((acc, arr) => { acc.push(...arr); return acc; }, []));
                if (nonFragments.length) {
                    const nonPartialNonFragments = nonFragments.filter(nf => !nf.isPartial);
                    const partialNonFragments = nonFragments.filter(nf => nf.isPartial);
                    if (nonPartialNonFragments.length) {
                        let builder = '';
                        builder += '{\n';
                        builder += nonPartialNonFragments.map(f => f.iface).join('\n');
                        builder += `\n${indentation}}`;
                        const newInterfaceName = generateSubTypeInterfaceName(selection.name.value, generatedTypeCount);
                        if (!newInterfaceName) {
                            andOps.push(builder);
                        }
                        else {
                            andOps.push(newInterfaceName);
                        }
                        generatedTypeCount += 1;
                        complexTypes.push({ iface: builder, isPartial: false, name: newInterfaceName });
                    }
                    if (partialNonFragments.length) {
                        let builder = '';
                        builder += 'Partial<{\n';
                        builder += partialNonFragments.map(f => f.iface).join('\n');
                        builder += `\n${indentation}}>`;
                        andOps.push(builder);
                        const newInterfaceName = generateSubTypeInterfaceName(selection.name.value, generatedTypeCount);
                        generatedTypeCount += 1;
                        complexTypes.push({ iface: builder, isPartial: true, name: newInterfaceName });
                    }
                }
                if (fragments.length) {
                    andOps.push(...fragments.map(wrapPartial));
                }
                const childType = andOps.join(' & ');
                str += convertToType(field.type, false, childType) + ';';
            }
            else {
                if (!field) {
                    console.log(selection);
                }
                str += convertToType(field.type) + ';';
            }
        }
        else if (selection.kind === 'FragmentSpread') {
            str = `IFragment${selection.name.value}`;
            isFragment = true;
            isPartial = isUndefinedFromDirective(selection.directives);
        }
        else if (selection.kind === 'InlineFragment') {
            const anon = !selection.typeCondition;
            if (!anon) {
                const typeName = selection.typeCondition.name.value;
                parent = parsedSchema.getType(typeName);
            }
            const selections = selection.selectionSet.selections.map(sel => getChildSelections(operation, sel, indentation, parent, !anon));
            let joinSelections = selections.map(s => s.iface).join('\n');
            isPartial = isUndefinedFromDirective(selection.directives);
            return {
                iface: joinSelections,
                isFragment,
                isPartial,
                complexTypes,
            };
        }
        else {
            throw new Error('Unsupported SelectionNode');
        }
        return {
            iface: str,
            isFragment,
            isPartial,
            complexTypes,
        };
    };
    const getVariables = variables => (variables.map(v => {
        const optional = v.type.kind !== 'NonNullType';
        return `${v.variable.name.value}${optional ? '?:' : ':'} ${convertVariable(v.type)};`;
    }));
    const variablesToInterface = (opName, variables) => {
        if (!variables || !variables.length) {
            return '';
        }
        const variableTypeDefs = getVariables(variables);
        return formatVariableInterface(opName, variableTypeDefs);
    };
    const buildAdditionalTypes = children => {
        const subTypes = children.reduce((acc, child) => { acc.push(...child.complexTypes); return acc; }, []);
        return subTypes.map(subtype => {
            if (subtype.isPartial) {
                return `export type ${subtype.name} = ${subtype.iface};`;
            }
            else {
                return `export interface ${subtype.name} ${subtype.iface}`;
            }
        });
    };
    return parsedSelection.definitions.map(def => {
        const ifaceName = buildRootInterfaceName(def);
        if (def.kind === 'OperationDefinition') {
            const variableInterface = variablesToInterface(ifaceName, def.variableDefinitions);
            const ret = def.selectionSet.selections.map(sel => getChildSelections(def.operation, sel, '  '));
            const fields = ret.map(x => x.iface);
            const iface = formatInterface(ifaceName, fields);
            const additionalTypes = buildAdditionalTypes(ret);
            return {
                variables: variableInterface,
                interface: iface,
                additionalTypes,
            };
        }
        else if (def.kind === 'FragmentDefinition') {
            // get the correct type
            const onType = def.typeCondition.name.value;
            const foundType = parsedSchema.getType(onType);
            const ret = def.selectionSet.selections.map(sel => getChildSelections('query', sel, '  ', foundType));
            const ext = ret.filter(x => x.isFragment).map(x => x.iface).join(', ');
            const extensions = ext ? ` extends ${ext}` : '';
            const fields = ret.filter(x => !x.isFragment).map(x => x.iface);
            const iface = formatFragmentInterface(ifaceName, fields, extensions);
            const additionalTypes = buildAdditionalTypes(ret);
            return {
                interface: iface,
                variables: '',
                additionalTypes,
            };
        }
        else {
            throw new Error(`Unsupported Definition ${def.kind}`);
        }
    });
};
exports.default = doIt;
//# sourceMappingURL=index.js.map