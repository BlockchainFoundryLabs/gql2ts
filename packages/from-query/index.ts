import {
  parse,
  GraphQLSchema,
  DocumentNode,
  GraphQLField,
  GraphQLCompositeType,
  isCompositeType,
  getNamedType,
  GraphQLType,
  VariableDefinitionNode,
  TypeNode,
  OperationTypeNode,
  GraphQLObjectType,
  GraphQLEnumType,
  DirectiveNode,
  GraphQLInputObjectType,
  GraphQLInputField,
  GraphQLUnionType,
  GraphQLNamedType,
  FieldNode,
} from 'graphql';
import {
  schemaFromInputs,
  isList,
  isNonNullable,
  isEnum,
} from '@gql2ts/util';
import {
  ChildSelectionsType,
  IChildren,
  IComplexTypeSignature,
  IOptions,
  Signature,
  RegularTypeSignature,
  VariableTypeSignature,
  convertToTypeSignature,
  ITypeMap,
} from '@gql2ts/types';
import {
  DEFAULT_TYPE_MAP,
  DEFAULT_OPTIONS,
} from '@gql2ts/language-typescript';
import {
  GenerateSubtypeCache,
  SubtypeNamerAndDedupe,
  ISubtypeMetadata,
} from './subtype';

const doIt: Signature = (schema, query, typeMap= {}, providedOptions= {}) => {
  const TypeMap: ITypeMap = {
    ...DEFAULT_TYPE_MAP,
    ...typeMap
  };

  const {
    buildRootInterfaceName,
    formatFragmentInterface,
    formatInterface,
    formatVariableInterface,
    wrapList,
    wrapPartial,
    generateSubTypeInterfaceName,
    printType,
    formatInput,
    generateFragmentName,
    generateQueryName,
    formatEnum,
    defaultIndentation,
    interfaceBuilder,
    typeBuilder,
    typeJoiner,
    generateInterfaceDeclaration,
    exportFunction,
  }: IOptions = { ...DEFAULT_OPTIONS, ...providedOptions };

  const getSubtype: SubtypeNamerAndDedupe = GenerateSubtypeCache();

  const parsedSchema: GraphQLSchema = schemaFromInputs(schema);
  const parsedSelection: DocumentNode = parse(query);

  const handleInputObject: (type: GraphQLInputObjectType, isNonNull: boolean) => string = (type, isNonNull) => {
    const variables: GraphQLInputField[] = Object.keys(type.getFields()).map(k => type.getFields()[k]);
    const variableDeclarations: string[] = variables.map(v => formatInput(v.name, true, convertToType(v.type)));
    const builder: string = generateInterfaceDeclaration(variableDeclarations.map(v => `    ${v}`), defaultIndentation);
    return printType(builder, isNonNull);
  };

  const handleEnum: (type: GraphQLEnumType, isNonNull: boolean) => string = (type, isNonNull) => {
    const decl: string = formatEnum(type.getValues());
    return printType(decl, isNonNull);
  };

  const handleNamedTypeInput: (type: TypeNode, isNonNull: boolean) => string | undefined = (type, isNonNull) => {
    if (type.kind === 'NamedType' && type.name.kind === 'Name' && type.name.value) {
      const newType: GraphQLType = parsedSchema.getType(type.name.value);
      if (newType instanceof GraphQLEnumType) {
        return handleEnum(newType, isNonNull);
      } else if (newType instanceof GraphQLInputObjectType) {
        return handleInputObject(newType, isNonNull);
      }
    }
  };

  const handleRegularType: RegularTypeSignature = (type, isNonNull, replacement) => {
    const typeValue: string = (typeof type.name === 'string') ? type.toString() : type.name.value;
    const showValue: string = replacement || typeValue;
    const show: string = TypeMap[showValue] || (replacement ? showValue : TypeMap.__DEFAULT);
    return printType(show, isNonNull);
  };

  const convertVariable: VariableTypeSignature = (type, isNonNull= false, replacement= null) => {
    if (type.kind === 'ListType') {
      return printType(wrapList(convertVariable(type.type, false, replacement)), isNonNull!);
    } else if (type.kind === 'NonNullType') {
      return convertVariable(type.type, true, replacement);
    } else {
      return handleNamedTypeInput(type, isNonNull!) || handleRegularType(type, isNonNull!, replacement!);
    }
  };

  const convertToType: convertToTypeSignature = (type, isNonNull= false, replacement= null): string => {
    if (isList(type)) {
      return printType(wrapList(convertToType(type.ofType, false, replacement)), isNonNull!);
    } else if (isNonNullable(type)) {
      return convertToType(type.ofType, true, replacement);
    } else if (isEnum(type)) {
      return handleEnum(type, isNonNull!);
    } else {
      return handleRegularType(type, isNonNull!, replacement!);
    }
  };

  const UndefinedDirectives: Set<string> = new Set(['include', 'skip']);

  const isUndefinedFromDirective: (directives: DirectiveNode[] | undefined) => boolean = directives => {
    if (!directives || !directives.length) { return false; }

    const badDirectives: DirectiveNode[] = directives.filter(d => !UndefinedDirectives.has(d.name.value));
    const hasDirectives: boolean = directives.some(d => UndefinedDirectives.has(d.name.value));

    if (badDirectives.length) {
      console.error('Found some unknown directives:');
      badDirectives.forEach(d => console.error(d.name.value));
    }

    return hasDirectives;
  };

  const getOperationFields: (operation: OperationTypeNode) => GraphQLObjectType = operation => {
    switch (operation) {
      case 'query':
        return parsedSchema.getQueryType();
      case 'mutation':
        return parsedSchema.getMutationType();
      case 'subscription':
        return parsedSchema.getSubscriptionType();
      default:
        throw new Error('Unsupported Operation');
    }
  };

  const wrapPossiblePartial: (possiblePartial: IChildren) => string = possiblePartial => {
    if (possiblePartial.isPartial) {
      return wrapPartial(possiblePartial.iface);
    } else {
      return possiblePartial.iface;
    }
  };

  const flattenComplexTypes: (children: IChildren[]) => IComplexTypeSignature[] = children => (
    children.reduce((acc, child) => { acc.push(...child.complexTypes); return acc; }, [] as IComplexTypeSignature[])
  );

  // @FIXME this is pretty bad
  const fixIndentationForSubtype: (decl: string) => string =
    decl => decl.replace(/\n\s+/g, `\n${defaultIndentation}`).replace(defaultIndentation + '}', '}');

  type GetField = (operation: OperationTypeNode, selection: FieldNode, parent?: GraphQLType) => GraphQLField<any, any>;
  const getField: GetField = (operation, selection, parent) => {
    if (parent && isCompositeType(parent)) {
      if (parent instanceof GraphQLUnionType) {
        return parent.getTypes().map(t => t.getFields()[selection.name.value]).find(z => !!z)!;
      } else {
        return parent.getFields()[selection.name.value];
      }
    } else {
      const operationFields: GraphQLObjectType = getOperationFields(operation);
      return operationFields.getFields()[selection.name.value];
    }
  };

  const getChildSelections: ChildSelectionsType = (operation, selection, indentation= '', parent?, isUndefined= false): IChildren => {
    let str: string = '';
    let isFragment: boolean = false;
    let isPartial: boolean = false;
    let complexTypes: IComplexTypeSignature[] = [];

    if (selection.kind === 'Field') {
      const field: GraphQLField<any, any> = getField(operation, selection, parent);
      const selectionName: string = selection.alias ? selection.alias.value : selection.name.value;
      let childType: string | undefined;

      isUndefined = isUndefined || isUndefinedFromDirective(selection.directives);
      let resolvedType: string;
      if (selectionName.startsWith('__')) {
        resolvedType = TypeMap.String;
      } else if (!!selection.selectionSet) {
        let newParent: GraphQLCompositeType | undefined;
        const fieldType: GraphQLNamedType = getNamedType(field.type);
        if (isCompositeType(fieldType)) {
          newParent = fieldType;
        }

        const selections: IChildren[] =
          selection.selectionSet.selections.map(sel => getChildSelections(operation, sel, indentation + defaultIndentation,  newParent));

        const nonFragments: IChildren[] = selections.filter(s => !s.isFragment);
        const fragments: IChildren[] = selections.filter(s => s.isFragment);
        const andOps: string[] = [];

        complexTypes.push(...flattenComplexTypes(selections));

        if (nonFragments.length) {
          const nonPartialNonFragments: IChildren[] = nonFragments.filter(nf => !nf.isPartial);
          const partialNonFragments: IChildren[] = nonFragments.filter(nf => nf.isPartial);

          if (nonPartialNonFragments.length) {
            const interfaceDeclaration: string = generateInterfaceDeclaration(nonPartialNonFragments.map(f => f.iface), indentation);
            const subtypeInfo: ISubtypeMetadata | null = getSubtype(selection, interfaceDeclaration, generateSubTypeInterfaceName);
            const newInterfaceName: string | null = subtypeInfo ? subtypeInfo.name : null;
            andOps.push(newInterfaceName || interfaceDeclaration);
            if (newInterfaceName && subtypeInfo && !subtypeInfo.dupe) {
              complexTypes.push({ iface: fixIndentationForSubtype(interfaceDeclaration), isPartial: false, name: newInterfaceName });
            }
          }

          if (partialNonFragments.length) {
            const interfaceDeclaration: string =
              wrapPartial(generateInterfaceDeclaration(partialNonFragments.map(f => f.iface), indentation));
            const subtypeInfo: ISubtypeMetadata | null = getSubtype(selection, interfaceDeclaration, generateSubTypeInterfaceName);
            const newInterfaceName: string | null = subtypeInfo ? subtypeInfo.name : null;
            andOps.push(newInterfaceName || interfaceDeclaration);
            if (newInterfaceName && subtypeInfo && !subtypeInfo.dupe) {
              complexTypes.push({ iface: fixIndentationForSubtype(interfaceDeclaration), isPartial: true, name: newInterfaceName });
            }
          }
        }

        andOps.push(...fragments.map(wrapPossiblePartial));
        childType = typeJoiner(andOps);
        resolvedType = convertToType(field.type, false, childType);
      } else {
        resolvedType = convertToType(field.type, false, childType);
      }
      str = formatInput(indentation + selectionName, isUndefined, resolvedType);
    } else if (selection.kind === 'FragmentSpread') {
      str = generateFragmentName(selection.name.value);
      isFragment = true;
      isPartial = isUndefinedFromDirective(selection.directives);
    } else if (selection.kind === 'InlineFragment') {
      const anon: boolean = !selection.typeCondition;
      if (!anon) {
        const typeName: string = selection.typeCondition!.name.value;
        parent = parsedSchema.getType(typeName);
      }

      const selections: IChildren[] =
        selection.selectionSet.selections.map(sel => getChildSelections(operation, sel, indentation, parent, !anon));

      let joinSelections: string = selections.map(s => s.iface).join('\n');
      isPartial = isUndefinedFromDirective(selection.directives);
      complexTypes.push(...flattenComplexTypes(selections));
      return {
        iface: joinSelections,
        isFragment,
        isPartial,
        complexTypes,
      };
    }

    return {
      iface: str,
      isFragment,
      isPartial,
      complexTypes,
    };
  };

  const getVariables: (variables: VariableDefinitionNode[]) => string[] = variables => (
    variables.map(v => {
      const optional: boolean = v.type.kind !== 'NonNullType';
      return formatInput(v.variable.name.value, optional, convertVariable(v.type));
    })
  );

  const variablesToInterface: (operationName: string, variables: VariableDefinitionNode[] | undefined) => string = (opName, variables) => {
    if (!variables || !variables.length) { return ''; }
    const variableTypeDefs: string[] = getVariables(variables);
    return exportFunction(formatVariableInterface(opName, variableTypeDefs));
  };

  const buildAdditionalTypes: (children: IChildren[]) => string[] = children => {
    const subTypes: IComplexTypeSignature[] = flattenComplexTypes(children);

    return subTypes.map(subtype => {
      if (subtype.isPartial) {
        return exportFunction(typeBuilder(subtype.name, subtype.iface));
      } else {
        return exportFunction(interfaceBuilder(subtype.name, subtype.iface));
      }
    });
  };

  return parsedSelection.definitions.map(def => {
    const ifaceName: string = buildRootInterfaceName(def, generateQueryName, generateFragmentName);
    if (def.kind === 'OperationDefinition') {
      const variableInterface: string = variablesToInterface(ifaceName, def.variableDefinitions);
      const ret: IChildren[] = def.selectionSet.selections.map(sel => getChildSelections(def.operation, sel, defaultIndentation));
      const fields: string[] = ret.map(x => x.iface);
      const iface: string = exportFunction(formatInterface(ifaceName, fields));
      const additionalTypes: string[] = buildAdditionalTypes(ret);

      return {
        variables: variableInterface,
        interface: iface,
        additionalTypes,
      };
    } else if (def.kind === 'FragmentDefinition') {
      // get the correct type
      const onType: string = def.typeCondition.name.value;
      const foundType: GraphQLType = parsedSchema.getType(onType);

      const ret: IChildren[] = def.selectionSet.selections.map(sel => getChildSelections('query', sel, defaultIndentation, foundType));
      const extensions: string[] = ret.filter(x => x.isFragment).map(x => x.iface);
      const fields: string[] = ret.filter(x => !x.isFragment).map(x => x.iface);
      const iface: string = exportFunction(formatFragmentInterface(ifaceName, fields, extensions));
      const additionalTypes: string[] = buildAdditionalTypes(ret);

      return {
        interface: iface,
        variables: '',
        additionalTypes,
      };
    } else {
      throw new Error(`Unsupported Definition ${def.kind}`);
    }
  });
};

export default doIt;
