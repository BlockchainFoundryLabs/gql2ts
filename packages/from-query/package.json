{
  "name": "@gql2ts/from-query",
  "private": false,
  "version": "0.6.0-alpha.a12f0f66",
  "description": "generate typescript interfaces from a graphql schema and query",
  "main": "index.js",
  "typings": "index.d.ts",
  "scripts": {
    "prepublish": "tsc -p ../../tsconfig.json"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/avantcredit/gql2ts.git"
  },
  "keywords": [
    "gql2ts",
    "graphql",
    "typescript"
  ],
  "author": "Brett Jurgens <brett.jurgens@avant.com> (https://www.avant.com)",
  "license": "ISC",
  "bugs": {
    "url": "https://github.com/avantcredit/gql2ts/issues"
  },
  "homepage": "https://github.com/avantcredit/gql2ts#readme",
  "files": [
    "*.js",
    "*.ts",
    "*.js.map"
  ],
  "dependencies": {
    "@gql2ts/language-typescript": "^0.6.0-alpha.a12f0f66",
    "@gql2ts/util": "^0.6.0-alpha.a12f0f66"
  },
  "devDependencies": {
    "@gql2ts/types": "^0.6.0-alpha.a12f0f66"
  }
}
