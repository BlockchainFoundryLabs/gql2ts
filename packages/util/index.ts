export { readFile, writeToFile } from './fileIO';
export { generateNamespace, writeNamespaceToFile } from './namespace';

export {
  PossibleIntrospectionInputs,
  PossibleSchemaInput,
  isIntrospectionResult,
  schemaFromInputs
} from './schema';
