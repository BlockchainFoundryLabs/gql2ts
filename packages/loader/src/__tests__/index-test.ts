import * as fs from 'fs';
import * as path from 'path';
import { Stats } from 'webpack';
import * as fromQuery from '@gql2ts/from-query';
import { PossibleSchemaInput } from '@gql2ts/util';
import { ITypeMap, IProvidedOptions } from '@gql2ts/types';
import compiler from './__helpers__/compiler';
import buildDeclaration from '../buildDeclaration';

const SOURCE: string = fs
  .readFileSync(path.resolve(__dirname, './graphql/query.graphql'))
  .toString();

describe('Loader', () => {
  beforeEach(() => {
    jest.restoreAllMocks();
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('fails if no schema is provided', async () => {
    const stats: Promise<Stats> = compiler(`./graphql/query.graphql`, {});
    await expect(stats).rejects.toThrow('Schema must be provided');
  });

  it('writes a declaration file', async () => {
    const RETURN_VALUE: string = 'GQL2TS RETURN VALUE';
    const FAKE_SCHEMA: PossibleSchemaInput = 'fake schema';

    const fsSpy: jest.Mock<typeof fs.writeFile> = jest
      .spyOn(fs, 'writeFile')
      .mockImplementationOnce((_file, data, callback) => {
        expect(data.toString()).toEqual(buildDeclaration(RETURN_VALUE));
        callback(null);
      });

    const gql2tsSpy: jest.Mock<typeof fromQuery.default> = jest
      .spyOn(fromQuery, 'default')
      .mockImplementationOnce(() => [{ result: RETURN_VALUE }]);

    await compiler(`./graphql/query.graphql`, {
      schema: FAKE_SCHEMA,
    });

    expect(fsSpy).toHaveBeenCalledTimes(1);
    expect(gql2tsSpy).toHaveBeenCalledTimes(1);
    expect(gql2tsSpy).toHaveBeenCalledWith(
      FAKE_SCHEMA,
      SOURCE,
      undefined,
      undefined,
    );
  });

  it('writes a declaration file using typeMap and options', async () => {
    const RETURN_VALUE: string = 'GQL2TS RETURN VALUE';
    const FAKE_SCHEMA: PossibleSchemaInput = 'fake schema';
    let typeMapValue: Partial<ITypeMap> = { Test: 'object', Test2: 'string' };
    let optionsValue: Partial<IProvidedOptions> = {
      wrapList: (type: string) => `wrapped(${type})`,
      generateSubTypeInterfaceName: () => null,
    };

    const fsSpy: jest.Mock<typeof fs.writeFile> = jest
      .spyOn(fs, 'writeFile')
      .mockImplementationOnce((_file, data, callback) => {
        expect(data.toString()).toEqual(buildDeclaration(RETURN_VALUE));
        callback(null);
      });

    const gql2tsSpy: jest.Mock<typeof fromQuery.default> = jest
      .spyOn(fromQuery, 'default')
      .mockImplementationOnce((_schema, _source, typeMap, options) => {
        expect(typeMap).toEqual(typeMapValue);
        expect(options).toEqual(optionsValue);
        return [{ result: RETURN_VALUE }];
      });

    await compiler(`./graphql/query.graphql`, {
      schema: FAKE_SCHEMA,
      typeMap: typeMapValue,
      options: optionsValue
    });

    expect(fsSpy).toHaveBeenCalledTimes(1);
    expect(gql2tsSpy).toHaveBeenCalledTimes(1);
    expect(gql2tsSpy).toHaveBeenCalledWith(
      FAKE_SCHEMA,
      SOURCE,
      typeMapValue,
      optionsValue,
    );
  });
});
