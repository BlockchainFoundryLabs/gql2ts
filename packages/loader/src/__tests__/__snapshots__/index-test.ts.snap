// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader fails if no schema is provided 1`] = `
[Error: Webpack Error: ../graphql/query.graphql
Module build failed (from /Users/bherma55/gql2ts/packages/loader/src/index.ts):
Error: Schema must be provided
    at Object.<anonymous>.module.exports (/Users/bherma55/gql2ts/packages/loader/src/index.ts:28:21)]
`;
